src/java.desktop/share/classes/java/awt/color/ICC_Profile.java

```
@@ -1131,17 +1146,36 @@ private static byte[] getData(Profile p, int tagSignature) {
      * This method is useful for advanced applications which need to access
      * profile data directly.
      *
+     * <p>
+     * Note: JDK built-in ICC Profiles cannot be updated using this method
+     * as it will result in {@code IllegalArgumentException}. JDK built-in
+     * profiles are those obtained by {@code ICC_Profile.getInstance(int colorSpaceID)}
+     * where {@code colorSpaceID} is one of the following:
+     * {@link ColorSpace#CS_sRGB}, {@link ColorSpace#CS_LINEAR_RGB},
+     * {@link ColorSpace#CS_PYCC}, {@link ColorSpace#CS_GRAY} or
+     * {@link ColorSpace#CS_CIEXYZ}.
+     * </p>
+     *
      * @param  tagSignature the ICC tag signature for the data element you want
      *         to set
      * @param  tagData the data to set for the specified tag signature
      * @throws IllegalArgumentException if {@code tagSignature} is not a
      *         signature as defined in the ICC specification.
-     * @throws IllegalArgumentException if a content of the {@code tagData}
+     * @throws IllegalArgumentException if the content of the {@code tagData}
      *         array can not be interpreted as valid tag data, corresponding to
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if this is a built-in profile for one
+     *         of the pre-defined ColorSpaces, i.e. those which can be obtained
+     *         by calling {@code ICC_Profile.getInstance(int colorSpaceID)}
```

can we just ignore it instead and did not use suspicion IllegalArgumentException for correct parameters? or change the type to something unrelated to "..ArgumentException"?

---

The silently do nothing option was considered, but if you do that, then you have no easy way of knowing if it worked.
Tests may pass spuriously, or fail later for the wrong reasons. So a worse choice.
And in all my searching of uses of this API it is (1) tests in the JDK itself and (2) a couple of libraries that are targeting specific known profiles with issues and are fixed up - so never applied to built-in profiles.

---

> can we just ignore it instead and did not use suspicion IllegalArgumentException for correct parameters? or change the type to something unrelated to "..ArgumentException"?

This has been [my suggestion](https://github.com/openjdk/jdk/pull/23606#discussion_r1968073992) since the start of this code review.

> The silently do nothing option was considered, but if you do that, then you have no easy way of knowing if it worked.  
> Tests may pass spuriously, or fail later for the wrong reasons. So a worse choice.

I agree, an exception should be thrown if modification isn't allowed.

> And in all my searching of uses of this API it is (1) tests in the JDK itself and (2) a couple of libraries that are targeting specific known profiles with issues and are fixed up - so never applied to built-in profiles.

Therefore we can throw <code>Illegal<i>State</i>Exception</code> instead of <code>Illegal<i>Argument</i>Exception</code>.

Quoting here [Harshitha's reply](https://github.com/openjdk/jdk/pull/23606#discussion_r1968263836):

> As Phil mentioned earlier, the already documented exception IAE for .setData() outweighs introducing a new exception.

Using the exception which clearly describes the reason why it was thrown outweighs re-using an already documented exception…

> Using an existing exception would mean less disruption to any existing applications using the Java API.

…especially because the search for usages of `setData` revealed that there are no usages of `setData` which will be affected by the proposed change.

If I'm not mistaken, the search also showed that no usages of `setData` handle `IllegalArgumentException` — thus apps will be equally unprepared to any type of an unchecked exception.

---

@aivanov-jdk @mrserb 
There are two other RTE types that may be better suited that ISE in this case - [UnsupportedOperationException ](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/UnsupportedOperationException.html) and [ProfileDataException](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/color/ProfileDataException.html) but the main motivation for going with IAE is that it is existing and documented exception.

> If I'm not mistaken, the search also showed that no usages of setData handle IllegalArgumentException — thus apps will be equally unprepared to any type of an unchecked exception.

Reasonable point.

---

I'm just not sure how this should be handled from the application's point of view. The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?

 - Clone the profile and then change the intent?
 - Change the intent, and if an exception occurs, clone the profile and try to change the intent again?
 - Skip changing the intent completely?

I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps. It wasn’t considered very important since the "appcontext was withdrawn". So, why are we deciding to change it now?

---

> I'm just not sure how this should be handled from the application's point of view. The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?
> 
> * Clone the profile and then change the intent?
> * Change the intent, and if an exception occurs, clone the profile and try to change the intent again?
> * Skip changing the intent completely?

If an app got a profile 'from some place', it probably should not be trying to change it, and likely has no idea if some isolated change will work properly, or the effect on what handed it that profile.
So copying the data and creating a new profile is what I'd recommend in such a case, regardless of this proposed restriction, and even then, only on a 'known' profile. 

> I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps. It wasn’t considered very important since the "appcontext was withdrawn". So, why are we deciding to change it now?

I don't see such a discussion there. Maybe it was 'offline' ?
Yes, this is less of an issue in a closed application, but it is an inappropriate thing to do/allow.

---

>If an app got a profile 'from some place', it probably should not be trying to change it,

But this is not always possible. For example, a buffered image has a color model that references a color space and a color profile. To modify something in that profile(like intents screen vs printer) before converting the entire image, it is necessary to check whether it is a built-in profile. If it is, the whole image must be recreated; otherwise, an exception will occur. Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).

>I don't see such a discussion there. Maybe it was 'offline' ?

It was somewhere online not too long ago.

---

>Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).

But it also has its own disadvantages...

---

The System Properties case isn't something I'd want to follow. It doesn't really make them read-only.
It just ensures that the JDK internally caches the original value for use in certain places.

Consider this program

public class jh {
  public static void main(String[] args) {
   String jhpstr = "java.home";
   String jhp1 = System.getProperty(jhpstr);
   System.setProperty(jhpstr, "foo");
   String jhp2 = System.getProperty("java.home");
   System.out.println(jhp1+" " +jhp2);
  }
} 
% java jh
/Library/Java/JavaVirtualMachines/jdk-21.jdk/Contents/Home foo

---

> There are two other exceptions that may be better suited that ISE in this case - [UnsupportedOperationException ](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/UnsupportedOperationException.html) and [ProfileDataException](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/color/ProfileDataException.html) but still IAE was chosen over them as it is existing and documented exception.

`UnsupportedOperationException` could be used.

`HeadlessException` extends `UnsupportedOperationException` and indicates a situation where a method cannot proceed because of the platform configuration.

`ProfileDataException` doesn't suit either because there's nothing wrong with the profile data in the case of built-in profiles.

---

> The System Properties case isn't something I'd want to follow. It doesn't really make them read-only.
> It just ensures that the JDK internally caches the original value for use in certain places.

I think you missed the main point. The application may attempt to change these java properties, but it won't affect the behavior of the JDK since the JDK internally will use the initially cached value of the property via StaticProperty.userDir()/userHome()/etc.

This is even specified similarly to what we want for color profiles.

>  * <strong>Changing a standard system property may have unpredictable results
     * unless otherwise specified.</strong>
     * Property values may be cached during initialization or on first use.
     * Setting a standard property after initialization using {@link #getProperties()},
     * {@link #setProperties(Properties)}, {@link #setProperty(String, String)}, or
     * {@link #clearProperty(String)} may not have the desired effect.

---

* Builtin color profiles may be cached during initialization or on first use. Changing a standard profiles after initialization using ICC_Profile.setData may not have the desired effect.

---

> > Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).
> 
> But it also has its own disadvantages...

@mrserb  You mentioned earlier that there are disadvantages of using java properties approach, so using builtIn flag would be better then, correct?

>  I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps.

I wasn't able to find the previous discussion thread so I'm not sure what are the disadvantages. Can you please explain?

---

If we start from the beginning. The problem we are trying to solve is preventing changes to our standard profiles so that we can be sure one part of the application does not unexpectedly affect another. This could be achieved in two ways:

- Throw an exception if a standard profile is modified
- Ignore the requested change and use the default profile properties as is

For new applications, we could require always cloning the color profile before modification (via the updated specification); otherwise, the application must always check whether the profile is built-in or not.

However, for older applications, these two solutions lead to different outcomes:

- Throwing the new exception will most likely break the application
- Ignoring the data change allows the application to continue working, but the color conversion result may not be fully accurate

The next question is: how big of an issue is this inaccuracy? For example, if we have an sRGB profile that rejects some modifications, we would still try to produce valid sRGB data during conversion.

So the trade-off is: broken applications vs. inaccurate (but still ICC-spec-compliant) conversions.

PS: By the way, why do we want to enforce this rule only for built-in profiles? Why can't a library create similar profiles and mark them as "read-only"?

---

@mrserb 

> The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?
> 
> Clone the profile and then change the intent?

The app would follow the above approach.
There is only one step extra with this fix - creating a copy of built-in profile and then modifying it.

```
byte[] builtInData = ICC_Profile.getInstance(ColorSpace.CS_sRGB).getData(); // get the byte array representation of BuiltIn- profile
ICCProfile newProfile = ICC_Profile.getInstance(builtInData) // create a new profile
newProfile.setData(...)
```


> However, for older applications, these two solutions lead to different outcomes:
> Throwing the new exception will most likely break the application

The chances of breaking an existing application is minimal since API usage for setData() was checked and not many were found and moreover they were not called on built-in profiles.


>Ignoring the data change allows the application to continue working, but the color conversion result may not be fully accurate

This does not seem viable because the user in unaware as to whether setData() worked or not.